### PR TITLE
Initial impl of 'convert anonymous type to tuple' refactoring

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Looks like this:

image

General idea is pretty simple.  you can start with an anonymous type and ask for it to be converted into a tuple.  If you do, then all matching anonymous type creations in that method will be converted.  This ensures that any relations between instances of the same anonymous type (i.e. assignments, collections, etc.) should be preserved.

This is part of a set of refactorings I'd like to provide around converting types around.  They include, but aren't limited to:

1. convert anonymous type to tuple (this pr).
2. convert anonymous type to named type.
3. convert tuple to named type.

I'm starting with anonymous types because they're conceptually simpler.  Because they cannot appear within signatures or 'leak' out of a method (in as far as the language is concerned), it makes it very easy to know what to update and what not to update.  Tuples are more complicated because it may be ambiguous whether two tuples with the same structure/type are actually intended by the user to represent the same entity.

I'm also intentionally *not* bothering with "convert tuple to anonymous type".  I just don't think it's important enough.

It may be interesting to offer 'convert named type to tuple' if the named type looks tuple-ish enough (i.e. no interesting methods on it, only state that is directly exposed, etc.)  But that's a but more complex as well, so i'm also holding off on that.

Todo:

- [x] VB support
- [x] Tests

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @kuhlenh @jinujoseph @sharwell @jcouv @heejaechang

**Comment by CyrusNajmabadi** [comment]
Also, thanks to @Neme12 for helping discuss this feature with me and to help come up with (and shoot down) some ideas :)

**Comment by heejaechang** [comment]
@dotnet/roslyn-ide is all code style analyzers on by default?

**Comment by Neme12** [comment]
I would worry about conversions to `object` and `dynamic` and not activate the analyzer in those cases (a refactoring would be fine though) since it's likely that there will be dynamic access or reflection used to get its properties (otherwise storing them as object would seem rather pointless).

For example ASP.NET Core uses anonymous types all over the place as a substitute for a dictionary:
image

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertAnonymousTypeToTuple
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpConvertAnonymousTypeToTupleCodeRefactoringProvider : AbstractCodeStyleDiagnosticAnalyzer
```
Comment (Neme12) [inline src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: the name seems rather fishy

File: src/Features/Core/Portable/Features.csproj [hunk 2]
```diff
@@ -110,5 +110,8 @@
   <ItemGroup>
     <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(SystemThreadingTasksExtensionsVersion)" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="ConvertAnonymousTypeToTuple\" />
```
Comment (Neme12) [inline src/Features/Core/Portable/Features.csproj]: is this needed?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Features.csproj]: i made the folder for the common impl, but haven't populated  it yet.  so this is waht VS does.

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: good catch. i flip-flopped on this being an analyzer/refactoring and didn't fix up the name.

**Comment by CyrusNajmabadi** [comment]
> @dotnet/roslyn-ide is all code style analyzers on by default?

Hey Heejae, can you clarify this question?  I'm not sure what it's asking.  Thanks!

**Comment by CyrusNajmabadi** [comment]
> I would worry about conversions to object and dynamic and not activate the analyzer in those cases

Definitely an interesting case.  I think i'll consider a warning in this case...

**Comment by Neme12** [comment]
I don't think a warning during the fix is good enough. The analyzer will still nag the user with bad suggestions all over the place, which will probably force them to disable this analyzer altogether, even though they would like it in cases where it *is* valid.

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        public CSharpConvertAnonymousTypeToTupleCodeRefactoringProvider()
+            : base(IDEDiagnosticIds.ConvertAnonymousTypeToTupleDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Convert_anonymous_type_to_tuple), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources.Convert_anonymous_type_to_tuple), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
```
Comment (jcouv) [inline src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: Does this analyzer need semantic information, or can it proceed just on syntax?

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                context.Diagnostics);
+
+            return SpecializedTasks.EmptyTask;
+        }
+
+        private async Task FixInCurrentMember(
+            Document document, SyntaxEditor editor,
+            Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            // For the standard invocation of the code-fix, we want to fixup all creations of the
```
Comment (jcouv) [inline src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs]: I don't know what happens with anonymous types in lambda or local functions. When are they considered "same" types (ie. to be replaced) or different?

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(FeaturesResources.Convert_anonymous_type_to_tuple, createChangedDocument, FeaturesResources.Convert_anonymous_type_to_tuple)
+            {
+            }
+        }
+    }
+}
```
Comment (jcouv) [inline src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs]: I took a quick look (looks good). I'll take a deeper look once the tests are added. VB seems fine to have its own PR.

**Comment by etbyrd** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
>  The analyzer will still nag the user with bad suggestions all over the place,

No, this is not a suggestion analyzer.  There is nothing noticeable to the user *except* that they will have a screwdriver available if their cursor is on the 'new' keyword.  

So there's no nagging and no suggestions.  This is just really a refactoring.  I happen to have used hte analyzer subsystem as it's the easy way to get behaviors i would otehrwise have to replicate if i was a refactoring.

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: Looks like it can just be syntax!

Comment (Neme12) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: not really... we should still protect against conversions to `object` and `dynamic`

**Comment by Neme12** [comment]
@CyrusNajmabadi I didn't notice that, thanks. I still believe there should be at least a warning in that case - but even that might not be enough when doing fix all. Imagine you want to do fix all in a project/solution that has a lot of places where this can be replaced with a tuple but just as many places where it cannot - the user will have to give up on doing "fix all" altogether, even though they would really like to do it in all the cases where there isn't a warning. I would prefer if we just filtered them out and didn't offer a fix there.

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: that's for the fix.  The analysis portion can be syntax only.

**Comment by CyrusNajmabadi** [comment]
> I would prefer if we just filtered them out and didn't offer a fix there.

I think that might be an acceptable approach.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide this is ready for review.  

Tagging @jcouv @sharwell @mavasani @heejaechang @jinujoseph

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 6]
```diff
@@ -1376,4 +1376,7 @@ This version used in: {2}</value>
   <data name="Convert_to_conditional_expression" xml:space="preserve">
     <value>Convert to conditional expression</value>
   </data>
+  <data name="Convert_anonymous_type_to_tuple" xml:space="preserve">
+    <value>Convert anonymous type to tuple</value>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FeaturesResources.resx]: We could also just do "Convert to tuple"

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            var anonymousType = (TAnonymousObjectCreationExpressionSyntax)context.Node;
+            if (GetInitializerCount(anonymousType) < 2)
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(
+                    Descriptor, context.Node.GetFirstToken().GetLocation(), ReportDiagnostic.Hidden,
+                    additionalLocations: null, properties: null));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: @sharwell This is your new DiagnosticHelper.Create form.  Can you let me know if i'm doing this properly?  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jcouv Could you look at this simple feature PR?  Thanks!

File: src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        {
+            var text = @"
+class Test
+{
+    void Method(int b)
+    {
+        var t1 = [||]new { a = 1, b = 2 };
+        var t2 = new { a = 3, b };
+        var t3 = new { a = 4 };
+        var t4 = new { b = 5, a = 6 };
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs]: >t4 [](start = 12, length = 2)

I was surprised the `t4` was left un-converted. It seems the same anonymous type as `t1`, no?

File: src/EditorFeatures/VisualBasicTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.vb [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        dim t1 = (a:=1, b:=2)
+        dim t2 = (a:=3, b:=4)
+    end sub
+end class
+"
+            Await TestInRegularAndScriptAsync(text, expected)
+        End Function
+
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertAnonymousTypeToTuple)>
+        Public Async Function ConvertMultipleInstancesAcrossMesthod() As Task
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.vb]: geez :D

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        TAnonymousObjectCreationExpressionSyntax> 
+        : SyntaxEditorBasedCodeFixProvider
+        where TExpressionSyntax : SyntaxNode
+        where TTupleExpressionSyntax : TExpressionSyntax
+        where TAnonymousObjectCreationExpressionSyntax : TExpressionSyntax
+    {
+        protected abstract TTupleExpressionSyntax ConvertToTuple(TAnonymousObjectCreationExpressionSyntax anonCreation);
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(IDEDiagnosticIds.ConvertAnonymousTypeToTupleDiagnosticId);
```
Comment (jcouv) [inline src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: Is this property called multiple times? Should we store the immutable array in a readonly static field?

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                   new LocalizableResourceString(nameof(FeaturesResources.Convert_to_tuple), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources.Convert_to_tuple), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        protected abstract TSyntaxKind GetAnonymousObjectCreationExpressionSyntaxKind();
+        protected abstract int GetInitializerCount(TAnonymousObjectCreationExpressionSyntax anonymousType);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
```
Comment (jcouv) [inline src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: >SemanticSpanAnalysis [](start = 42, length = 20)

Can this be a syntax analyzer only (no semantics)?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: most likely!  thanks!

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(c => FixAllWithEditorAsync(context.Document,
+                    e => FixInCurrentMember(context.Document, e, context.Diagnostics[0], c), c)),
+                context.Diagnostics);
+
+            return SpecializedTasks.EmptyTask;
```
Comment (jcouv) [inline src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: >EmptyTask [](start = 36, length = 9)

I'm not sure why we have this static value. Task.CompletedTask is shorter and seems clearer.

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        private async Task FixInCurrentMember(
+            Document document, SyntaxEditor editor,
+            Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            // For the standard invocation of the code-fix, we want to fixup all creations of the
+            // "same" anonymous type within the containing method.  We define same-ness as meaning
+            // "they have the type symbol".  this means both have the same member names, in the same
```
Comment (jcouv) [inline src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: >this [](start = 45, length = 4)

Typo: "this" should be upper-cased

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: I think this predates Task.CompletedTask.  And SpecializedTasks.EmptyTask just defers to Task.CopmletedTask.  I'm happy to make another PR that removes the old form and updates everyone.

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+            // order, with the same member types.  We fix all these up in the method because the
+            // user may be creating several instances of this anonymous type in that method and
+            // then combining them in interesting ways (i.e. checking them for equality, using them
+            // in collections, etc.).  The language guarantees within a method boundary that these
+            // will be the same type and can be used together in this fashion.
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```
Comment (jcouv) [inline src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: This local can be moved closer to usage.

File: src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+using Microsoft.CodeAnalysis.CSharp.ConvertAnonymousTypeToTuple;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertAnonymousTypeToTuple
+{
+    public partial class ConvertAnonymousTypeToTupleTests : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
+    {
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs]: >{ [](start = 4, length = 1)

Would you mind testing in various containers? (local function, lambda)

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs (hunk 4 above)
Comment (jcouv) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs]: I didn't hear back from you on this question

---
In reply to: [190789240](https://github.com/dotnet/roslyn/pull/27099#discussion_r190789240) [](ancestors = 190789240)

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs]: I went with: 

"they should be viewed as the same type, and we'd want to change them all from the highest method's span inwards."

However, i didn't do tests.  I will add them now :)

File: src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs (hunk 15 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs]: Will do.

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: Will do.

File: src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs (hunk 15 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs]: Maybe also test on incomplete code `new { a = , b = }`?

---
In reply to: [197975162](https://github.com/dotnet/roslyn/pull/27099#discussion_r197975162) [](ancestors = 197975162)

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: I have opened a PR to remove that old type entirely :)

File: src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConvertAnonymousTypeToTuple/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs]: Done

File: src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs (hunk 15 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToTuple/ConvertAnonymousTypeToTupleTests.cs]: tests added (for both C# and VB).

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleCodeFixProvider.cs]: Done.

**Comment by CyrusNajmabadi** [comment]
@jcouv Tests added.  Rereview at your convenience.

**Comment by CyrusNajmabadi** [comment]
@sharwell @mavasani @heejaechang @ivanbasov can any of you review?  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jinujoseph for merging approval.  @kuhlenh For new feature notification.

**Comment by CyrusNajmabadi** [comment]
@kuhlenh also note the sibling feature PR: https://github.com/dotnet/roslyn/pull/28066

**Comment by CyrusNajmabadi** [comment]
thanks  @jcouv @ivanbasov

**Comment by CyrusNajmabadi** [comment]
@jinujoseph Great.  Then both this and https://github.com/dotnet/roslyn/pull/28066 have been reviewed and are ready to go in.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jinujoseph This has passed tests and has been reviewed.  It is ready to go in.  Thanks!

**Comment by CyrusNajmabadi** [comment]
test windows_debug_vs-integration_prtest

**Comment by CyrusNajmabadi** [comment]
test windows_release_vs-integration_prtest

**Comment by CyrusNajmabadi** [comment]
test windows_debug_vs-integration_prtest

**Comment by CyrusNajmabadi** [comment]
test windows_release_vs-integration_prtest

**Comment by CyrusNajmabadi** [comment]
@jcouv This has been reviewed, and everythign passes.  can you merge in?  Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**